syntax = "proto3";

package helenite.core;

import "google/protobuf/wrappers.proto";
import "google/protobuf/empty.proto";

service Master {
  rpc CreateFile(CreateFileRequest) returns (google.protobuf.BoolValue) {}
  rpc AllocateChunk(AllocateChunkRequest) returns (ChunkInformation) {}
  rpc GetChunkInformation(ChunkHandle) returns (ChunkInformation) {}
  rpc RegisterChunkServer(ChunkServerAddress) returns (google.protobuf.BoolValue) {}
  rpc UnregisterChunkServer(ChunkServerAddress) returns (google.protobuf.Empty) {}
}

service ChunkServer {
  rpc Heartbeat (google.protobuf.Empty) returns (google.protobuf.Empty) {}
}

message CreateFileRequest {
  string filename = 1;
}

message AllocateChunkRequest {
  string filename = 1;
  int64 sequenceNumber = 2;
}

message BulkAllocateChunkRequest {
  string filename = 1;
  int64 count = 2;
}

message ChunkHandle {
  string handle = 1;
}

message ChunkInformation {
  ChunkHandle handle = 1;
  repeated ChunkServerAddress servers = 3;
}

message ChunkServerAddress {
  string address = 1;
}